pkg/ipcache/metadata.go

```diff
@@ -0,0 +1,378 @@
+					"released":       released,
+					logfields.Labels: lbls,
+				}).Debug(
+					"Releasing old identity with previously assigned to IP and updating with new set of labels",
+				)
+			}
+		}
+	}
+
+	// If no other labels are associated with this IP, we assume that it's
+	// outside of the cluster and hence needs a CIDR identity. This might be a
+	// temporary identity.
+	if lbls.Equals(labels.LabelKubeAPIServer) {
+		// The release of the identitiy allocated has been handled above. This
+		// can happen if we discover that the prefix is also associated with
+		// other labels besides kube-apiserver, i.e. if the kube-apiserver is
+		// not running outside of the cluster.
+		return injectLabelsForCIDR(prefix, lbls)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.IPAllocationTimeout)
+	defer cancel()
+	return IdentityAllocator.AllocateIdentity(ctx, lbls, false)
+}
+
+// injectLabelsForCIDR will allocate a CIDR identity for the given prefix. The
+// release of the identity must be managed by the caller.
+func injectLabelsForCIDR(prefix string, lbls labels.Labels) (*identity.Identity, bool, error) {
+	if !strings.Contains(prefix, "/") {
+		prefix = prefix + "/32"
+	}
+	_, cidr, err := net.ParseCIDR(prefix)
+	if err != nil {
+		return nil, false, err
+	}
+
+	allLbls := cidrlabels.GetCIDRLabels(cidr)
+	allLbls.MergeLabels(lbls)
+
+	log.WithFields(logrus.Fields{
+		logfields.CIDR:   cidr,
+		logfields.Labels: lbls, // omitting allLbls as CIDR labels would make this massive
+	}).Debug(
+		"Injecting CIDR labels for prefix",
+	)
+
+	return allocate(cidr, allLbls)
+}
+
+// FilterMetadataByLabels returns all the prefixes inside the IdentityMetadata
+// map which contain the given labels. Note that `filter` is a subset match,
+// not a full match.
+func FilterMetadataByLabels(filter labels.Labels) []string {
+	idMDMU.RLock()
+	defer idMDMU.RUnlock()
+	matching := make([]string, 0)
+	for prefix, lbls := range IdentityMetadata {
+		if bytes.Contains(lbls.SortedList(), filter.SortedList()) {
+			matching = append(matching, prefix)
+		}
+	}
+	return matching
+}
+
+// RemoveAllPrefixesWithLabels wraps RemoveLabels to provide a convenient
+// method for the caller to remove all given prefixes at once. This function
+// will trigger policy update and recalculation if necessary on behalf of the
+// caller if any changes to the kube-apiserver were detected.
+//
+// Identities allocated by InjectLabels() may be released by RemoveLabels().
+//
+// A prefix will only be removed from the IDMD if the set of labels becomes
+// empty.
+func RemoveAllPrefixesWithLabels(
+	m map[string]labels.Labels,
+	src source.Source,
+	updater identityUpdater,
+	triggerer policyTriggerer,
+) {
+	var (
+		hasKubeAPIServer bool
+
+		idsToPropagate = make(map[identity.NumericIdentity]labels.LabelArray)
+	)
+	for prefix, lbls := range m {
+		id, exists := IPIdentityCache.LookupByIP(prefix)
+		if id.ID <= identity.IdentityUnknown || !exists {
```

Comment:
The `IdentityUnknown` case here looks like a no-op, ipcache should never return `IdentityUnknown` and it's 0 so < should never trigger. Not sure if you were trying to catch another case here but if so we'll need to revisit that.

Reply:
Hm, I don't recall anything specific that I saw. I think I was just being extra defensive.